brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1118 @@
+
+                param0_fitV[:n_V - 1] = res_fitV.x.copy()
+
+                # fit U, the covariance matrix, together with AR(1) param
+                param0_fitU[0:n_l] = current_vec_U_chlsk_l_AR1
+                param0_fitU[n_l:n_l + n_V] = current_a1
+                res_fitU = scipy.optimize.minimize(
+                    self._loglike_y_AR1_diagV_fitU, param0_fitU,
+                    args=(XTX, XTDX, XTFX, YTY_diag, YTDY_diag, YTFY_diag,
+                          XTY, XTDY, XTFY, current_logSNR2, l_idx, n_C,
+                          n_T, n_V, rank),
+                    method='BFGS', jac=True, tol=10**-2,
+                    options={'xtol': 10**-3, 'disp': self.verbose,
+                             'maxiter': 3})
+                current_vec_U_chlsk_l_AR1 = res_fitU.x[0:n_l]
+                current_a1 = res_fitU.x[n_l:n_l + n_V]
+                norm_fitUchange = np.linalg.norm(res_fitU.x - param0_fitU)
+                logger.debug('norm of parameter change after fitting U: '
+                             '{}'.format(norm_fitUchange))
+                param0_fitU = res_fitU.x.copy()
+
+                # if norm_fitVchange / np.sqrt(param0_fitV.size - n_smooth)\
+                #         < tolx and norm_fitUchange\
+                #         / np.sqrt(param0_fitU.size) < tolx:
+                #     break
+
+            current_GP[0] = np.log(np.min(
+                dist2[np.tril_indices_from(dist2, k=-1)]))
+#             current_GP[0] = np.log(np.max(dist2)/4.0)
+            # Initialize it with the smallest distance.
+            # Alternatively, initialize with a large distance.
+            # In testing, the result does depend on the initialization.
+            # There can be two local minimum when the data does not
+            # constrain the spatial landscape of log(SNR2) enough
+            # I think this is because of the addition of epsilon.
+            if GP_inten:
+                current_GP[1] = np.log(np.maximum(
+                    np.percentile(inten_diff2[np.tril_indices_from(
+                        inten_diff2, k=-1)], 5), 0.5))
+
+            param0_fitV[n_V - 1:n_V - 1 + n_smooth] = current_GP
+            logger.debug('current GP[0]:{}'.format(current_GP[0]))
+
+        tolx = self.tolx
+        n_iter = self.n_iter
+        logger.debug('start real fitting')
+        for it in range(0, n_iter):
+
+            # fit V
+
+            res_fitV = scipy.optimize.minimize(
+                self._loglike_y_AR1_diagV_fitV, param0_fitV, args=(
+                    XTX, XTDX, XTFX, YTY_diag, YTDY_diag, YTFY_diag, XTY, XTDY,
+                    XTFY, current_vec_U_chlsk_l_AR1, current_a1, l_idx,
+                    n_C, n_T, n_V, rank, GP_space, GP_inten, dist2,
+                    inten_diff2, space_smooth_range, inten_smooth_range,
+                    inten_weight),
+                method='BFGS', jac=True, tol=10**(-2 - 2 / n_iter * (it + 1)),
```

Comment:
Tighter tolerance with increasing iterations is good -- but it's hard to figure out what the decay schedule is. Would be good to comment (and also respect tolx if it is being passed in).